File: CSC330/collatzBad/julia/collatz.jl
#!/usr/bin/julia

# Count sequence in Collatz Conjecture iteratively
function CountCollatz(n)
      counter = 0
      while n > 1
            if(n % 2 == 0)
                  n = n / 2
            else
                  n = n * 3 + 1
            end
            counter += 1
      end
      return counter
end;

# Will find the smallest sequence length and replace it with a number's current calculated sequence length
function ReplaceLeast(sequences, nums, seqCount, currentNum)

      # If the current sequence count is lower than the smallest sequence stored, then it
      # can't be bigger than the rest. So skip and return slice
      if(sequences[1] > seqCount)
            return sequences, nums
      end
      
      # Otherwise, check all the elements in the array
      for i in eachindex(sequences)
            
            if(sequences[i] == seqCount) 
                  return sequences, nums
            end
            
            if(sequences[i] > seqCount) 
                  sequences, nums = ShiftLeft(sequences, nums, i - 1, seqCount, currentNum)
                  return sequences, nums
            end
        
      end
      
      # Otherwise, the current sequence count is larger than all the stored sequences,
      # so the index before must be the spot for the current sequence count
      sequences, nums = ShiftLeft(sequences, nums, 10, seqCount, currentNum)
      return sequences, nums
         
end

# The function will shift the element in array to the left by 1
# It helps the ReplaceLeast function replace the elements accordingly with the new values
function ShiftLeft(sequences, nums, pos, seq, currentNum)

      newSeq = Array{Int64}(undef, 10)
      newNums = Array{Int64}(undef, 10)
      
      # The top (pos + 1) indicies will be the same as original array
      for i = 10:-1:pos+1
            newSeq[i] = sequences[i]
            newNums[i] = nums[i]
      end
      
      # Set the new elements at the desired pos
      newSeq[pos] = seq
      newNums[pos] = currentNum
      
      # Shift elements from indicies pos below to the new arrays
      for i = pos - 1:-1:1
            newSeq[i] = sequences[i+1]
            newNums[i] = nums[i+1]
            #println(i)
      end
      
      # Return the new shifted arrays
      return newSeq, newNums
      
end

# Prints two arrays reverse order
function PrintArrays(sequences, nums)

      for i = 10:-1:1
            println(nums[i], "       ", sequences[i])
      end

end

# Main
function main()

      print("\nEnter a number greater than 10 : ")
      r = parse(Int64, readline())
      println()

      # First 10 sequences are set by default
      sequences = [0, 1, 2, 3, 5, 6, 7, 8, 16, 19]
      nums = [1, 2, 4, 8, 5, 10, 3, 6, 7, 9]

      # Find top 10 smallest integers with the largest sequences
      for i in 11:r
            sequences, nums = ReplaceLeast(sequences, nums, CountCollatz(i), i)
      end
      
      println("Sorted based on sequence length")
      PrintArrays(sequences, nums)
      
      # Sort nums and sequences so that num is ordered from least to greatest
      # sequences will be sorted accordingly with nums
      p = sortperm(nums); nums .= nums[p]; sequences .= sequences[p]
      
      println("Sorted based on integer size")
      PrintArrays(sequences, nums)
      
      println()
      
end

main()

File: CSC330/collatz/go/collatz.go
package main

import ("fmt"
        "sort"
)

// Struct can be used to hold a number and its corresponding Collatz Sequence length
type sequence struct { 
      num int
      sequences int
}

func main() {

      var r int
      fmt.Print("Enter a number greater than 10 : ")
      fmt.Scan(&r)
      FindTop10(r)
      
}

// Finds the top 10 largest sequences with the smallest integers
func FindTop10(r int) { 

      numSeq := make([]sequence, 10)
      
      // Fill up the first 10 slots with sequence lengths for 1 - 10
      for i := 0; i < 10; i++ { 
            numSeq[i].sequences = CountCollatz(i + 1)
            numSeq[i].num = i + 1
      }
      
      // Sort array of structs by sequence length from least to greatest
      sort.Slice(numSeq, func(i, j int) bool {
            return numSeq[i].sequences < numSeq[j].sequences
      })
      
      // Find top 10 
      for i := 10; i < r; i++ { 
            numSeq = ReplaceLeast(numSeq, CountCollatz(i+1), i+1)
      }
      
      // Prints nums and sequence length from greatest sequence length to least greatest sequence length
      fmt.Println("Sorted based on sequence length")  
      PrintArray(numSeq)
      
      // Sort array of structs by integer length from least to greatest
      sort.Slice(numSeq, func(i, j int) bool {
            return numSeq[i].num < numSeq[j].num
      })     
      
      // Prints nums and sequence length from greatest integer to least greatest integer
      fmt.Println("Sorted based on integer size")
      PrintArray(numSeq)

}

// Counts the sequence length of a number iteratively
func CountCollatz(n int) int{ 
    
    var counter int = 0
    for n > 1 { 
          if(n % 2 == 0) { 
                n = n / 2
          } else  { 
                n = n * 3 + 1
          }
          counter++
    }
    return counter
    
}

// Will find the smallest sequence length and replace it with a number's current calculated sequence length
func ReplaceLeast(slice []sequence, seqCount int, currentNum int) []sequence { 
      
      // If the current sequence count is lower than the smallest sequence stored, then it
      // can't be bigger than the rest. So skip and return slice
      if(slice[0].sequences > seqCount) {
            return slice
      }
      
      // Otherwise, check all the elements in the array
      for i := 0; i < len(slice); i++ {
      
            // The currentNum will ALWAYS be bigger than the stored num, so skip and return
            if(slice[i].sequences == seqCount) { 
                  return slice
            }

            // If here, and the sequences stored is greater than the current sequences, then
            // the index before must be the spot for the current sequence count
            if(slice[i].sequences >= seqCount) {
                  slice = ShiftLeft(slice, i - 1, seqCount, currentNum)
                  return slice
            }
                  
      }
      
      // Otherwise, the current sequence count is larger than all the stored sequences,
      // so replace the last index and shift other elements down accordingly
      slice = ShiftLeft(slice, len(slice) - 1, seqCount, currentNum)
      return slice
      
}

// The function will shift the element in array to the left by 1
// It helps the ReplaceLeast function replace the elements accordingly with the new values
func ShiftLeft(array []sequence, pos int, seq int, currentNum int) []sequence { 

      demo := make([]sequence, len(array))
      for i := len(array) - 1; i > pos; i-- { 
            demo[i].sequences = array[i].sequences
            demo[i].num = array[i].num
      }
      demo[pos].sequences = seq
      demo[pos].num = currentNum
      for i := pos - 1; i >= 0; i-- { 
            demo[i].sequences = array[i+1].sequences
            demo[i].num = array[i+1].num
      }
      return demo
}

// Prints an array reverse order
func PrintArray(numSeq []sequence) { 

      for i := len(numSeq) - 1; i >= 0; i-- {
            fmt.Println(numSeq[i].num, "   ", numSeq[i].sequences)
      }

}
